Civi/Webshoptools/HandleWebShopActivity.php

```diff
@@ -0,0 +1,165 @@
+      $data['custom_field_data'] = $customField;
+      $data['custom_field_id'] = $customField['id'];
+      $data['entity_field_name'] = 'custom_' . $customField['id'];
+      $data['old_value'] = empty($currentActivity[$data['entity_field_name']]) ? '' : $currentActivity[$data['entity_field_name']];
+
+      if (empty($params['custom'][$customField['id']])) {
+        $data['new_value'] = '';
+        $data['new_value_label'] = '';
+      } else {
+        foreach ($params['custom'][$customField['id']] as $field) {
+          if (!empty($field['id'])) {
+            $fieldRowId = $field['id'];
+          }
+          $data['new_value'] = $field['value'];
+
+          if (!empty($customField['option_group_id']) && !empty($data['new_value'])) {
+            $data['new_value_label'] = civicrm_api3('OptionValue', 'getvalue', [
+              'return' => 'label',
+              'option_group_id' => $customField['option_group_id'],
+              'value' => $data['new_value'],
+            ]);
+          } else {
+            $data['new_value_label'] = $field['value'];
+          }
+        }
+      }
+
+      $preparedCustomFields[$customField['name']] = $data;
+    }
+
+    foreach ($customFields['values'] as $customField) {
+      $preparedCustomFields[$customField['name']]['params_key'] = $preparedCustomFields[$customField['name']]['entity_field_name'] . '_' . $fieldRowId;
+      $preparedCustomFields[$customField['name']]['field_row_id'] = (int) $fieldRowId;
+    }
+
+    return $preparedCustomFields;
+  }
+
+  /**
+   * @return string
+   */
+  private static function generateSubject($customFieldsData) {
+    $orderType = $shirtType = $shirtSize = $orderCount = '';
+
+    if (!empty($customFieldsData['order_type']['new_value_label'])) {
+      $orderType = $customFieldsData['order_type']['new_value_label'] . ' ';
+    }
+
+    if (!empty($customFieldsData['shirt_type']['new_value_label']) && !empty($customFieldsData['shirt_size']['new_value_label'])) {
+      $shirtType = $customFieldsData['shirt_type']['new_value_label'] . '/';
+    } elseif (!empty($customFieldsData['shirt_type']['new_value_label'])) {
+      $shirtType = $customFieldsData['shirt_type']['new_value_label'] . ' ';
+    }
+
+    if (!empty($customFieldsData['shirt_size']['new_value_label'])) {
+      $shirtSize = $customFieldsData['shirt_size']['new_value_label'] . ' ';
+    }
+
+    if (!empty($customFieldsData['order_count']['new_value_label'])) {
+      $orderCount = $customFieldsData['order_count']['new_value_label'] . 'x';
+    }
+
+    return trim($orderType . $shirtType . $shirtSize . $orderCount);
+  }
+
+  /**
+   * @param $currentActivity
+   * @param $params
+   * @param $operation
+   * @return bool
+   */
+  private static function isNewActivityStatusCompleted($currentActivity, $params, $operation) {
+    if ($operation != 'edit') {
+      return false;
+    }
+
+    $currentStatusId = $currentActivity['status_id'];
+    $newStatusId = $params['status_id'];
```

Comment:
`$params['status_id']` may be empty under some circumstances, e.g. in the `CRM_Webshoptools_WebshoptoolsTest::testActivitySubjectIsCorrect()` unit test.